lib/src/arithmetic/dotproduct.dart

```diff
@@ -0,0 +1,126 @@
+// Copyright (C) 2025 Intel Corporation
+// SPDX-License-Identifier: BSD-3-Clause
+//
+// dotproduct.dart
+// An integer dot-product unit.
+//
+// 2025 July 23
+// Author: Desmond A Kirkpatrick <desmond.a.kirkpatrick@intel.com>
+
+import 'package:collection/collection.dart';
+import 'package:rohd/rohd.dart';
+import 'package:rohd_hcl/rohd_hcl.dart';
+
+/// A dot product module for integers.
+class DotProduct extends Module {
+  /// The dot-product result [product].
+  Logic get product => output('product');
+
+  /// Creates a new [DotProduct] instance given a [List<Logic>] of
+  /// [multiplicands] and a [List<Logic>] of [multipliers], computing their
+  /// dot-product.  Currently widths of all operands must match.
+  ///
+  /// The [radix] parameter specifies the radix for use in partial-product
+  /// generation of the multiplies. While a [ColumnCompressor] is used on the
+  /// tall array of partial products, the final addition is accomplished using
+  /// the specified [adderGen] (default is [NativeAdder.new]).
+  ///
+  /// The optional [signedMultiplicand] parameter configures the [multiplicands]
+  /// statically using a `bool` to indicate a signed multiplicand (default is
+  /// `false`, or unsigned) or dynamically with a 1-bit [Logic] input. Passing
+  /// something other null, `bool`, or [Logic] will result in a throw.
+  ///
+  ///
+  /// The optional [signedMultiplier] parameter configures the [multipliers]
+  /// statically using a `bool` to indicate a signed multiplier (default is
+  /// `false`, or unsigned) or dynamically with a 1-bit [Logic] input.  Passing
+  /// something other null, `bool`, or [Logic] will result in a throw.
+  ///
+  /// The output [product] will be [log2Ceil(multiplicands.length)] wider than
+  /// the sum of the widths of one pair of products to accomadate the increasing
+  /// accumulation value.
+  DotProduct(List<Logic> multiplicands, List<Logic> multipliers,
+      {int radix = 4,
+      dynamic signedMultiplicand,
+      dynamic signedMultiplier,
+      Adder Function(Logic a, Logic b, {Logic? carryIn}) adderGen =
+          NativeAdder.new,
+      super.name = 'dotproduct',
+      super.reserveName = false,
+      super.reserveDefinitionName = false,
+      String? definitionName})
+      : super(
+            definitionName:
+                definitionName ?? 'DotProduct_W{$multipliers[0].width}_') {
+    if (multipliers.length != multiplicands.length) {
+      throw RohdHclException(
+          'Number of multipliers and multiplicands must be equal.');
+    }
+    final candWidthMiss = multiplicands
```

Comment:
I don't understand how this `candWidthMiss` uses `length` to determine the index of the first (or any) index where the widths mismatched?

Reply:
`candWidthMiss` produces a vector of `true`s whenever the `multiplicand` width matches the `multiplier` width.  If that vector is less than the `multiplicands` length, then it is a failure.  The `where` trims the list to just `true`s.  

Unfortunately it counts the mismatches rather than stopping at the first, so I'll adjust the message for now.

I'll add a test to ensure it throws.

Reply:
ah yeah your change makes sense to me now :)